### ISSUE TITLE
[BUG] Issues with device atomic overloads for wrapper types

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#1398]
**Describe the bug**
The device atomic overloads for the wrapper types hardcode the underlying type of the wrappers. This is brittle as it will break if the underlying types ever change. Furthermore, it's using machine dependent types, e.g., `long long int` instead of `int64_t`.

See: 
https://github.com/rapidsai/cudf/blob/b63e971f9c2348db6c12882c882139b2e0988d1e/cpp/src/utilities/device_atomics.cuh#L560

Furthermore, why does `cudf::date32 atomicAdd(cudf::date32* address, cudf::date32 val)` invoke `typesAtomicOperation32` but `cudf::date64 atomicAdd(cudf::date64* address, cudf::date64 val)` invokes `genericAtomicOperation`. It seems like there's an opportunity for consolidating functionality there. 

In general, I think an atomic overload for a wrapper type should look something like this:

```
inline  __device__
cudf::date32 atomicAdd(cudf::date32* address, cudf::date32 val)
{
    return cudf::date32{ atomicAdd( static_cast< cudf::detail::unwrapped_type_t<cudf::date32>* >(address),
                                                        unwrap(val)) };
 }
```

Unless I'm missing something, I don't think there's a need for the `typesAtomicOperation32/64`.

**Comment by kovaltan** [comment]
@jrhemstad 
> The device atomic overloads for the wrapper types hardcode the underlying type of the wrappers. This is brittle as it will break if the underlying types ever change.

Agree. That's a issue.

> Furthermore, it's using machine dependent types, e.g., long long int instead of int64_t.

Actually there is an issue between long long int and int64_t.
For example,
When I changed T_int from `unsigned long long int` into `uint64_t` at https://github.com/rapidsai/cudf/blob/b63e971f9c2348db6c12882c882139b2e0988d1e/cpp/src/utilities/device_atomics.cuh#L187-L197
 
nvcc causes errors with below message while `atomicCAS` are defined for `unsigned long long int`
Is there any way to fix it?

> cudf/cpp/src/utilities/device_atomics.cuh(192): error: no instance of overloaded function "atomicCAS" matches the argument list
>             argument types are: (uint64_t *, uint64_t, uint64_t)
>           detected during instantiation of "T cudf::detail::typesAtomicCASImpl<T, 8UL>::operator()(T *, const T &, const T &) [with T=int64_t]"


> Furthermore, why does cudf::date32 atomicAdd(cudf::date32* address, cudf::date32 val) invoke typesAtomicOperation32 but cudf::date64 atomicAdd(cudf::date64* address, cudf::date64 val) invokes genericAtomicOperation. It seems like there's an opportunity for consolidating functionality there.

Cuda's `atomicAdd` supports signed 32bit, but not signed 64bit. That's the reason why date32 and date64 has different implementation.
In order to consolidate functionality, if `genericAtomicOperation` is specialized to use native atomics like I did in PL #1005, all functions can call `genericAtomicOperation`, it would work and we can remove `typesAtomicOperation32/64`

**Comment by jrhemstad** [comment]
> nvcc causes errors with below message while atomicCAS are defined for unsigned long long int
Is there any way to fix it?

Hm, that is very annoying. I guess you'd need to provide `atomic*` functions for the fixed width types, e.g., `int64_t`, that cast to the type that CUDA supports. Just to be extra safe, I'd add a `static_cast(sizeof(int64_t) == sizeof(long long int))`. 

> In order to consolidate functionality, if `genericAtomicOperation` is specialized to use native atomics like I did in PL #1005, all functions can call `genericAtomicOperation`, it would work and we can remove `typesAtomicOperation32/64`

I think this is a good solution.

**Comment by kovaltan** [comment]
Also PR #1142 adds cudf::bool8 type.
In the PR, we discussed we should move `atomicCASImpl(int8 or int16)` into `typesAtomicCASImpl`.

**Comment by harrism** [comment]
I also would like the fix for this issue to enable atomics for cudf::bool8. I think doing that is outside the scope of PR #1142 .

**Comment by kovaltan** [comment]
Closed since PR #1691 fixed the issue.

**Comment by harrism** [comment]
#1735 was merged, but into branch-0.7.2... Now that that is merged there needs to be a merge of master into branch-0.8 to get those fixes in 0.8...

**Comment by kovaltan** [comment]
Oops, I was supposed that #1691 was merged into branch 0.8, but actually it was just closed but not merged yet.
Reopening the issue for issue tracking.

**Comment by harrism** [comment]
This is now merged into 0.8 so closing.
